src/transformers/sagemaker/trainer_sm.py

```diff
@@ -0,0 +1,177 @@
+from ..trainer_pt_utils import (
+    DistributedLengthGroupedSampler,
+    SequentialDistributedSampler,
+    nested_detach,
+    nested_numpify,
+)
+from ..utils import logging
+from .training_args_sm import is_smdistributed_available
+
+
+logger = logging.get_logger(__name__)
+
+
+if is_smdistributed_available():
+    import smdistributed.modelparallel.torch as smp
+
+    @smp.step()
+    def forward_backward(model, inputs, gradient_accumulation_steps=1):
+        outputs = model(**inputs)
+        loss = outputs["loss"] if isinstance(outputs, dict) else outputs[0]
+        loss = loss / gradient_accumulation_steps
+        model.backward(loss)
+        return loss
+
+    @smp.step()
+    def forward_only(model, inputs):
+        return model(**inputs)
+
+    def smp_gather(tensor):
+        if isinstance(tensor, (list, tuple)):
+            return type(tensor)(smp_gather(t) for t in tensor)
+        elif isinstance(tensor, dict):
+            return type(tensor)({k: smp_gather(v) for k, v in tensor.items()})
+        elif not isinstance(tensor, torch.Tensor):
+            raise TypeError(
+                f"Can't gather the values of type {type(tensor)}, only of nested list/tuple/dicts of tensors."
+            )
+        all_tensors = smp.allgather(tensor, smp.CommGroup.DP_GROUP)
+        return torch.cat([t.cpu() for t in all_tensors], dim=0)
+
+    def nested_smp_concat(tensor):
+        if isinstance(tensor, (list, tuple)):
+            return type(tensor)(nested_smp_concat(t) for t in tensor)
+        elif isinstance(tensor, dict):
+            return type(tensor)({k: nested_smp_concat(v) for k, v in tensor.items()})
+        # It doesn't seem possible to check here if `tensor` is a StepOutput because StepOutput lives in `smp.step`
+        # which is also the name of the decorator so Python is confused.
+        return tensor.concat().detach().cpu()
+
+
+class SageMakerTrainer(Trainer):
+    def __init__(self, args=None, **kwargs):
+        super().__init__(args=args, **kwargs)
+        self.is_model_parallel_enabled = is_smdistributed_available() and self.args.mp_parameters != ""
+
+    def _get_train_sampler(self):
+        if self.is_model_parallel_enabled:
+            if self.args.group_by_length:
+                return DistributedLengthGroupedSampler(
+                    self.train_dataset, self.args.train_batch_size, num_replicas=smp.dp_size(), rank=smp.dp_rank()
+                )
+            else:
+                return DistributedSampler(self.train_dataset, num_replicas=smp.dp_size(), rank=smp.dp_rank())
+        else:
+            return super()._get_train_sampler()
+
+    def _get_eval_sampler(self, eval_dataset: Dataset) -> Optional[torch.utils.data.sampler.Sampler]:
+        if self.is_model_parallel_enabled:
+            return SequentialDistributedSampler(eval_dataset, num_replicas=smp.dp_size(), rank=smp.dp_rank())
+        else:
+            return super()._get_eval_sampler(eval_dataset)
```

Comment:
Same behavior as regular `Trainer` but we need to `smp.dp_size()` and `smp.dp_rank()` for the data parallelism (some processes use the same data if they are on the model parallelism axis).

Reply:
When `is_model_parallel_enabled` and iterable datset is used, this will give error that iterable dataset does not have lenght because distributed sampler queries for it. We should also handle iterable dataset case separately here as done in main trainer class.

Reply:
Iterable dataset in a distributed training is not really supported in the regular Trainer (in the sense that all the processes will see the same data) so I'm fine with an error being raised here, since this is a distributed setup.

Reply:
(Of course when we roll out support for iterable datasets for distributed training, we will definitely add a special case here.)

Reply:
@sgugger  For Iterable dataset, logic to make sure that all processes don’t see  the same data needs to be implemented in the dataset. The main trainer class still supports iterable dataset - https://github.com/huggingface/transformers/blob/1c8c2d9ab34b8c8d326db9e0608f8e54cfccb885/src/transformers/trainer.py#L446.  Perhaps we can  handle it similarly here and print a warning to take care of data duplication in dataset?